pipeline_server/src/main.rs

```diff
@@ -0,0 +1,873 @@
+//!
+//! # API concepts
+//!
+//! * Project.  A project is a SQL script with a non-unique name and a unique ID
+//!   attached to it.  The client can add/remove/modify/compile a project.
+//!   Compilation includes running the SQL-to-DBSP compiler followed by the Rust
+//!   compiler.
+//!
+//! * Configuration.  A project can have multiple configurations associated with
+//!   it.  A configuration specifies input and output streams as a YAML file
+//!   that is deserialized into a `ControllerConfig` object.  Similar to
+//!   projects, one can add/remove/modify configs.
+//!
+//! * Pipeline.  A pipeline is a running instance of a compiled project.  One
+//!   can start multiple pipelines for a project with the same or different
+//!   configs.
+//!
+//! The API is currently single-tenant: there is no concept of users or
+//! permissions.  Multi-tenancy can be implemented by creating a server
+//! instance per tenant, which enables better separation of concerns,
+//! resource isolation and fault isolation compared to buiding multitenancy
+//! into the server.
+//!
+//! # Architecture
+//!
+//! * Project database.  Projects (including SQL source code), configs, and
+//!   pipelines are stored in a Postgres database.  The database is the only
+//!   state that is expected to survive across server restarts.  Intermediate
+//!   artifacts stored in the file system (see below) can be safely deleted.
+//!
+//! * Compiler.  The compiler generates a binary crate for each project and adds
+//!   it to a cargo workspace that also includes libraries that come with the
+//!   SQL libraries.  This way, all precompiled dependencies of the main crate
+//!   are reused across projects, thus speeding up compilation.
+//!
+//! * Runner.  The runner component is responsible for starting and killing
+//!   compiled pipelines and for interacting with them at runtime.  It also
+//!   registers each pipeline with Prometheus.
+//!
+//! # Concurrency
+//!
+//! We want to prevent race conditions due to users accessing the same project
+//! from different browsers.  An example is user 1 modifying the project,
+//! while user 2 is starting a pipeline for the same project.  The pipeline
+//! may end up running the old or the new version.  Our very simple solution
+//! is to increment project version on each update.  Every request to compile
+//! the project or start a pipeline must include project id _and_ version
+//! number. If the version number isn't equal to the current version in the
+//! database, this means that the last version of the project observed by the
+//! user is outdated, so the request is rejected.  Note that we don't store old
+//! versions, only the latest one.
+
+// TODOs:
+// * Tests.
+// * Generate a proper JSON API + docs using, e.g., OpenAPI.
+// * Support multi-node DBSP deployments (the current architecture assumes that
+//   pipelines and the Prometheus server run on the same host as this server).
+// * Proper UI.
```

Comment:
this todo does not belong to this file

Reply:
This is crate-level documentation, and the UI is part of the crate.